Managed implementation of SetAllVector128 SSE2 helper HW intrinsic

4creators opened the pull request:
FYI @CarolEidt @fiigii @mikedn @tannergooding

mikedn commented:
Have you checked what code this approach generates? Just out of curiosity, I don't think it matters too much at this point. This should be good enough for 2.1, Improvements can be done later if needed.

4creators commented:
Yes, it is similar to this what was generated when everything was expanded in importer except for (U)Int32 and (U)Int64 which are almost perfect ... quad word has wrong instruction
 
- double
```assembly
       C4E14857F6           vxorps   xmm6, xmm6

G_M22785_IG12:
       C4E17157C9           vxorpd   xmm1, xmm1, xmm1
       C4E17310CE           vmovsd   xmm1, xmm1, xmm6
       C4E171C6C100         vshufpd  xmm0, xmm1, xmm1, 0
```

- quad word sized operands
```assembly
       C4E1796EC6           vmovd    xmm0, xrsi
       C4E17970C044         vpshufd  xmm0, xmm0, 68
```

double word sized operands
```assembly
       C4E1796EC6           vmovd    xmm0, xrsi
       C4E17970C000         vpshufd  xmm0, xmm0, 0
```

- word sized operands:
```assembly
       0FB7CE               movzx    rcx, si
       C4E1796EC1           vmovd    xmm0, xrcx
       C4E17961C8           vpunpcklwd xmm1, xmm0, xmm0
       C4E17970C900         vpshufd  xmm1, xmm1, 0
```

- byte sized operands:

```assembly
       400FB6CE             movzx    rcx, sil
       C4E1796EC1           vmovd    xmm0, xrcx
       C4E17960C8           vpunpcklbw xmm1, xmm0, xmm0
       C4E17828C1           vmovaps  xmm0, xmm1
       C4E17960C8           vpunpcklbw xmm1, xmm0, xmm0
       C4E17828C1           vmovaps  xmm0, xmm1
       C4E17970C000         vpshufd  xmm0, xmm0, 0
```

tannergooding commented:
I had expected a change in the importer, that returned `nullptr` for these nodes, so that we wouldn't hit an `unreached`.

However, looking at the (existing) code, this looks to be "falling out" because we have `switch(isa) { default: return nullptr; }`, so any intrinsics not in the table will go here... I'm not sure this is the correct thing to be doing. CC. @CarolEidt

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -1023,47 +1024,90 @@ public static class Sse2
+        {
+            Vector128<byte> vector = Sse.StaticCast<uint, byte>(ConvertScalarToVector128UInt32(value));
+            vector = UnpackLow(vector, vector);
+            return Sse.StaticCast<uint, byte>(Shuffle(Sse.StaticCast<byte, uint>(UnpackLow(vector, vector)), 0));
```
nit: These lines are hard to read.

We should introduce temps or indent these lines, or something. @CarolEidt or @eerhardt might be able to point to the "proper" formatting for this kind of code.

4creators replied:
The problem here is that temps tend to be followed by additional XMM register allocated in jit, my preference would be to avoid them and try to express it better with proper indentation.

4creators replied:
In the case of `byte` it is visible as multiple additional unnecessary `movaps xmm1, xmm0`

tannergooding replied:
@CarolEidt, does the register allocator not recognize and optimize away temps that are write-once, read-multiple?

CarolEidt replied:
> does the register allocator not recognize and optimize away temps that are write-once, read-multiple?

I'm not sure what you're asking here - if it is read-multiple it can't be optimized away. Indeed, if it's used at all, it can't really be "optimized away". However, note that all the interim values here will require allocation, and may be spilled to temps anyway.
If it is write-once, read-once (as I think you would get from the above), it should *theoretically* behave the same in the register allocator, but won't always be due to differences in the order of the trees in the IR.
In any case, we should write the code cleanly, and file an issue (*with* a buildable test case) when there are unnecessary moves.

tannergooding commented:
For `byte` and `sbyte`:
* I had expected `punpcklbw` followed by `punpcklwd`
* Note for future: we only need one xmm register for the whole thing, hopefully we can get the register allocator to cooperate
  * We likely want to fix this in general, since this is the type of code actual users may write as well

For `short` and `ushort`:
* Looks good

For `int` and `uint`:
* Looks good

For `long` and `ulong`:
* Had expected `movq` instead of `movd`
* Had expected `punpcklqdq` instead of `pshufd`

For `double`:
* Not sure why we have a `xorps` and a `xorpd`, we should probably be testing non-zero values
  * Is this a bug in `SetScalarVector128` or just a side-effect of the input values?
* Had expected `unpcklpd` instead of `shufpd`

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -1023,47 +1024,90 @@ public static class Sse2
+        public static Vector128<double> SetAllVector128(double value)
+        {
+            Vector128<double> vector = SetScalarVector128(value);
+            return Shuffle(vector, vector, 0);
```
`UnpackLow` will result in slightly smaller codegen and will do the same thing (instructions are listed at same latency/throughput).

4creators replied:
Will update

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -1023,47 +1024,90 @@ public static class Sse2
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<ulong> SetAllVector128(ulong value)
+        {
+            return Sse.StaticCast<int, ulong>(Shuffle(Sse.StaticCast<ulong, int>(ConvertScalarToVector128UInt64(value)), (byte)0b01000100));
```
`UnpackLow` (for long/ulong) will do the same thing as `Shuffle (68)`, but with slightly smaller codegen (instructions are listed at same latency/throughput).

4creators replied:
Will update

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -1023,47 +1024,90 @@ public static class Sse2
+        public static Vector128<byte> SetAllVector128(byte value)
+        {
+            Vector128<byte> vector = Sse.StaticCast<uint, byte>(ConvertScalarToVector128UInt32(value));
+            vector = UnpackLow(vector, vector);
```
This should be `Vector128<ushort> tmp = Sse.StaticCast<byte, ushort>(UnpackLow(vector, vector))`, because the second `UnpackLow` needs to generate `punpcklwd`

4creators replied:
Right, my error.

tannergooding commented:
@4creators, It looks like `ConvertScalarToVector128Int64` is doing the wrong thing because it is using `INS_mov_i2xmm` but without passing along the proper EA_*BYTE size (this looks to be the same reason why some instructions print the second argument wrong `movd     xmm0, xrcx`)

For x86, it needs to be using `INS_movq` and reading the value from memory (since it can't load it from register) -- This change is more complicated and can probably wait.

4creators commented:
> It looks like ConvertScalarToVector128Int64 is doing the wrong thing because it is using INS_mov_i2xmm

Right, it seems to be culprit for `movd` instead of `movq`

4creators commented:
> Note for future: we only need one xmm register for the whole thing

Yes, I know it very well, unfortunately we fighting here with jit.

4creators commented:
By changing for `long` and `ulong`  `pshufd` to `punpcklqdq` we get:

```assembly
       66480F6EC6           movd     xmm0, rsi
       0F28C8               movaps   xmm1, xmm0
       660F6CC8             punpcklqdq xmm1, xmm0
```

What is not what we want - `pshufpd` is much better just bcs it takes one vector argument.

4creators commented:
For `byte` and `sbyte` disasm is OK after change with exception for many `movaps xmm0, xmm1` and usage of 2 registers instead of one but it similar to previous result:

```assembly
       480FBECE             movsx    rcx, sil
       660F6EC1             movd     xmm0, xrcx
       0F28C8               movaps   xmm1, xmm0
       660F60C8             punpcklbw xmm1, xmm0
       0F28C1               movaps   xmm0, xmm1
       0F28C8               movaps   xmm1, xmm0
       660F61C8             punpcklwd xmm1, xmm0
       660F70C900           pshufd   xmm1, xmm1, 0
```

tannergooding commented:
> By changing for long and ulong pshufd to punpcklqdq we get

This looks to be an issue with the register allocator not saying that the target and source register are the same. With the unneeded copy removed, we would generate:
```
66 48 0f 6e c6          movq   xmm0,rsi
66 0f 6c c0             punpcklqdq xmm0,xmm0
```

Which is one byte shorter than using `shufpd` (also assuming "optimal" codegen):
```
66 48 0f 6e c6          movq   xmm0,rsi
66 0f c6 c0 00          shufpd xmm0,xmm0,0x0
```

4creators commented:
> This looks to be an issue with the register allocator not saying that the target and source register are the same

Yep, perhaps we should keep shuffle for now and later work on solving this issues in jit. Anyway, I would be surprised if everything would work perfectly at this stage of development.

Going to log issue for this.

tannergooding commented:
> Yep, perhaps we should keep shuffle for now and later work on solving this issues in jit.

I think we should still keep `UnpackLow` and just aim to make the register allocator smarter (if that is possible). `Shuffle` probably only "looks" better because for this particular piece of code it decided that target/src could be the same. I doubt that it will always do that (but haven't actually looked if that is true), in which case the resulting code will still be sometimes larger and sometimes smaller.

4creators commented:
> Shuffle probably only "looks" better

I don't think so, there is a reason why native compilers prefer shuffle to unpack. Perhaps it is making register allocator work easier.

4creators commented:
> probably only "looks" better because for this particular piece of code it decided that target/src could be the same

IMO this bcs shuffle takes one argument while unpack takes two. It will by design require more nursing to get "the good" codegen.

tannergooding commented:
@4creators, the native compilers don't, they prefer unpack: https://godbolt.org/g/dFX6uj

4creators commented:
> the native compilers don't, they prefer unpack: https://godbolt.org/g/dFX6uj

IMO online compilers/decompilers test sites are not that reliable. I always prefer to do some dirty job myself and what I have got from MSVC is as follows:

- byte
```assembly
00007FF7911C1012  movsx       eax,byte ptr [valueSt (07FF7911C3034h)]  
00007FF7911C1019  movd        xmm0,eax  
00007FF7911C101D  punpcklbw   xmm0,xmm0  
00007FF7911C1021  punpcklwd   xmm0,xmm0  
00007FF7911C1025  pshufd      xmm0,xmm0,0
```
- short
```assembly
00007FF61A551012  movsx       eax,word ptr [valueSt (07FF61A553034h)]  
00007FF61A551019  movd        xmm0,eax  
00007FF61A55101D  punpcklwd   xmm0,xmm0  
00007FF61A551021  pshufd      xmm0,xmm0,0
```
- int
```assembly
00007FF667DB1012  movd        xmm0,dword ptr [valueSt (07FF667DB3034h)]  
00007FF667DB101A  pshufd      xmm0,xmm0,0
```
- long
```
00007FF6FC791012  movsxd      rax,dword ptr [valueSt (07FF6FC793034h)]  
00007FF6FC791019  movq        xmm0,rax  
00007FF6FC79101E  punpcklqdq  xmm0,xmm0
```

tannergooding commented:
> IMO online compilers/decompilers test sites are not that reliable

Not sure why, they are literally compiling the code and returning the generated assembly, same as if you had done it locally.

> what I have got from MSVC is as follows

That is exactly what the compiler site returns, the only differences are coming from you not specifying `__vectorcall` (which is how the code will generally behave as when/if inlined -- provided stack spilling hasn't happened) and from (it looks like) you compiling for x86 (rather than x86-64).

It is also worthwhile to look at the output of multiple compilers. Each has strengths/weaknesses and I am not aware of one that always produces the most optimal code.

For `SetAllVector128(long)`, both GCC and MSVC are using `punpcklqdq`, Clang is the only one using `pshufd`.

mikedn commented:
While it's likely possible to improve register allocation for cases such as shuffle(x, x) it may be useful to consider adding overloads that take a single parameter so people can simply write shuffle(x). Assembly instructions have 2 sources so they can provided maximum flexibility but in a high level language such flexibility can turn into a nuisance. And can result in worse code generation if the compiler isn't good enough at optimizing things.

4creators commented:
> you compiling for x86 (rather than x86-64)

I compiled for both x86 and x64

fiigii commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse.cs:
```diff
@@ -479,7 +480,12 @@ public static class Sse
+        {
+            Vector128<float> vector = SetScalarVector128(value);
+            return Shuffle(vector, vector, 0);
+        }
```
Could you try to optimize `SetAllVector128` via `Avx2.BroadcastScalarToVector128<T>`  to see whether the `if` block impacts inlining?
For example,
```csharp
        public static Vector128<float> SetAllVector128(float value)
        {
            if (Avx2.IsSupported)
            {
                return Avx2.BroadcastScalarToVector128<float>(value);
             }
             else
             {
                Vector128<float> vector = SetScalarVector128(value);
                return Shuffle(vector, vector, 0);
             }
        }
```

fiigii replied:
Actually, the final solution should be:
```csharp
        public static Vector128<float> SetAllVector128(float value)
        {
             if (Avx2.IsSupported)
             {
                return Avx2.BroadcastScalarToVector128<float>(value);
             }
             else if (Sse41.IsSupported)
             {
                // use SSE4.1 insert
             }
             else
             {
                Vector128<float> vector = SetScalarVector128(value);
                return Shuffle(vector, vector, 0);
             }
        }
```
So I worry about that the inlining with the code size increasing.

tannergooding replied:
@AndyAyersMS, this type of code might fail the initial inlining check, but should be marked as inlineable after the dead code is dropped (all of the if-blocks not selected, since `IsSupported` is a constant), correct?

4creators replied:
I will test it shortly

AndyAyersMS replied:
Not quite sure what you're asking -- the method likely will not be inlined without the aggressive inlining attribute. The inliner can't easily anticipate that the method body will shrink down once we've realized the calls therein expand to jit intrinsics with known values.

Generally we don't mark things as inlineable -- we reject them as possible candidates and possibly mark them as never inlinable to speed up future method jitting. The rejection happens in stages -- an initial pass in the importer and then a second pass in the inliner proper. Once a call is rejected it is never reconsidered, and (aside from the never inline stuff) we don't remember anything from past attempts, either when jitting this method or jitting any other.

4creators replied:
With brodcast the disasm is as follows:

```assembly
       C4E17A1005B6010000   vmovss   xmm0, dword ptr [reloc @RWD00]
       C4E17A1145AC         vmovss   dword ptr [rbp-54H], xmm0
       488D4DAC             lea      rcx, bword ptr [rbp-54H]
       C4E2791801           vbroadcastss xmm0, xmmword ptr [rcx]
```

We have again `vmovss` doing circles between stack and XMM but otherwise method is inlined and expanded as expected.

4creators replied:
Method implementation is:

```C#
        /// <summary>
        /// __m128 _mm_set1_ps (float a)
        ///   HELPER
        /// </summary>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static Vector128<float> SetAllVector128(float value)
        {
            if (Avx2.IsSupported)
            {
                 unsafe { return Avx.BroadcastScalarToVector128(&value); }
            }
            else
            {
                Vector128<float> vector = SetScalarVector128(value);
                return Shuffle(vector, vector, 0);
            }
        }
```

fiigii replied:
No, we should use Avx2.BroadcastScalarToVector128 rather than Avx.

4creators replied:
@AndyAyersMS What do you think about having besides `AggressiveInlining` an `AlwaysInlining` value in `MethodImpOpt`. Would it be hard to get it implemented in jit?

AndyAyersMS replied:
`AggressiveInlining` is already as aggressive as we can get -- if the jit can inline, it will. We are not holding anything back. The dial is already turned up to 11.

There are necessary limitations.

Some methods can't ever be inlined, or can't currently be inlined. And we pass on inlining in some rare situations -- for instance you can't use `AggressiveInlining` and recursion to create unboundedly large methods.

4creators replied:
> No, we should use Avx2.BroadcastScalarToVector128 rather than Avx.

`Avx2.BroadcastScalarToVector128` accepts `Vector128<float>` not float therefore signature in your example is invalid.

fiigii replied:
> Avx2.BroadcastScalarToVector128 accepts Vector128<float> not float

Oh, I see. So that we also need `SetScalarVector128`, and we cannot use `Avx.BroadcastScalarToVector128` because storing arguments into memory and reading it again is really slow.
Looks like one more reason to implement these helpers in codgen.

CarolEidt commented:
> However, looking at the (existing) code, this looks to be "falling out" because we have switch(isa) { default: return nullptr; }, so any intrinsics not in the table will go here... I'm not sure this is the correct thing to be doing.

I don't think it's unreasonable for the default to return nullptr, but it would probably be cleaner to have an assert as the default (still returning nullptr for safety), and enumerate the cases where we recognize the intrinsic, but are returning nullptr.

4creators commented:
> but it would probably be cleaner to have an assert as the default (still returning nullptr for safety), and enumerate the cases where we recognize the intrinsic, but are returning nullptr.

Will do

4creators commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

4creators commented:
Expecting failures on Sse42.Crc32 tests

4creators commented:
Addressed code review feedback. Changes in jit implementation of handling of `managed` intrinsics require merging #16817 first 

@CarolEidt @fiigii @tannergooding PTAL

4creators commented on src/jit/hwintrinsichelperlistxarch.h:
```diff
@@ -0,0 +1,32 @@
+/* Note
+1) Each hardware intrinsic has a unique Intrinsic ID with type of `enum NamedIntrinsic`
+2) All the overloads of an intrinsic in an ISA class share one Intrinsic ID
+3) The helper intrinsic which is implemented in software with use of hardware intrinsics and is part of public API is defined in this table
```
Intrinsics which were not in the table with managed implementation were handled by jit as a normal function which due to AggresiveInlining attribute was inlined and intrinsics used in implementation were expanded.

Currently if any of the static classes implement managed intrinsics they have to be included in this table. To make it easier to include native and managed intrinsics tables they were separated into two files.

4creators commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -139,7 +145,15 @@ NamedIntrinsic Compiler::lookupHWIntrinsic(const char* methodName, InstructionSe
+            return NI_Illegal;
+        }
+
+        assert((result > NI_HW_INTRINSIC_START) && (result < NI_SW_INTRINSIC_START));
```
This assert will be hit by any function present in ISA static classes and not in one of the tables preventing execution of HW intrinsic APIs  which are not supported by jit.

tannergooding replied:
I think the assert should be `result < NI_HW_INTRINSIC_END`

4creators replied:
With above change it will fir even better.

4creators commented:
All test failures are due to #16848

tannergooding commented:
@4creators, could you rebase and retest, so that tests will pass?

4creators commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

tannergooding commented on src/jit/hwintrinsichelperlistxarch.h:
```diff
@@ -0,0 +1,32 @@
+//                                                                                                                                                                            {TYP_BYTE,      TYP_UBYTE,       TYP_SHORT,     TYP_USHORT,    TYP_INT,       TYP_UINT,      TYP_LONG,      TYP_ULONG,     TYP_FLOAT,     TYP_DOUBLE}
+// ************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************ ...
+//  SSE Intrinsics
+SOFTWARE_INTRINSIC(SSE_SetAllVector128,                              "SetAllVector128",                                  SSE,        -1,           16,          1,            {INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,      INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid},        HW_Category_Helper,                            HW_Flag_NoFlag)
```
You could probably define a helper macro that doesn't require you to fill out all 10 instruction fields (which will always be invalid).

4creators replied:
If `SOFTWARE_INTRINSIC` macro does not define all `HWIntrinsicInfo` members compiler raises the following warnings and errors:

```
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(25): warning C4003: not enough actual parameters for macro 'SOFTWARE_INTRINSIC' (compiling source file E:\src\ms\dotnet\coreclr-sse2\
src\jit\hwintrinsicxarch.cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(26): warning C4003: not enough actual parameters for macro 'SOFTWARE_INTRINSIC' (compiling source file E:\src\ms\dotnet\coreclr-sse2\
src\jit\hwintrinsicxarch.cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(28): warning C4003: not enough actual parameters for macro 'SOFTWARE_INTRINSIC' (compiling source file E:\src\ms\dotnet\coreclr-sse2\
src\jit\hwintrinsicxarch.cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(29): warning C4003: not enough actual parameters for macro 'SOFTWARE_INTRINSIC' (compiling source file E:\src\ms\dotnet\coreclr-sse2\
src\jit\hwintrinsicxarch.cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\VC\VCTargets\Microsoft.CppCommon.targets(171,5): warning MSB4220: Waiting for the currently executing task "CustomBuild" to ca
ncel. [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\inc\corguids.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(25): error C2220: warning treated as error - no 'object' file generated (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\
hwintrinsicxarch.cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(25): error C2059: syntax error: ',' (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsicxarch.cpp) [E:\src\ms\dot
net\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(25): error C2143: syntax error: missing ';' before '}' (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsicxarch.
cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(25): error C2059: syntax error: '}' (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsicxarch.cpp) [E:\src\ms\dot
net\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(26): error C2143: syntax error: missing ';' before '{' (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsicxarch.
cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(26): error C2447: '{': missing function header (old-style formal list?) (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\
hwintrinsicxarch.cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(26): error C2059: syntax error: ',' (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsicxarch.cpp) [E:\src\ms\dot
net\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(28): error C2143: syntax error: missing ';' before '{' (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsicxarch.
cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(28): error C2447: '{': missing function header (old-style formal list?) (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\
hwintrinsicxarch.cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(28): error C2059: syntax error: ',' (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsicxarch.cpp) [E:\src\ms\dot
net\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(29): error C2143: syntax error: missing ';' before '{' (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsicxarch.
cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(29): error C2447: '{': missing function header (old-style formal list?) (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\
hwintrinsicxarch.cpp) [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsichelperlistxarch.h(29): error C2059: syntax error: ',' (compiling source file E:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsicxarch.cpp) [E:\src\ms\dot
net\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsicxarch.cpp(29): error C2143: syntax error: missing ';' before '}' [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrji
t.vcxproj]
e:\src\ms\dotnet\coreclr-sse2\src\jit\hwintrinsicxarch.cpp(29): error C2059: syntax error: '}' [E:\src\ms\dotnet\coreclr-sse2\bin\obj\Windows_NT.x64.Checked\src\jit\standalone\clrjit.vcxproj]
```

tannergooding replied:
Yes, which is why you define it as:
#define SOFTWARE_INTRINSIC(x, y, z) HARDWARE_INTRINSIC(x, y, INS_invalid, INS_invalid, z)

You then don't need to define the 10 instruction entries (which are always INS_invalid for software intrinsics), but will still have all the required HWIntrinsic Info fields defined

tannergooding commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -139,7 +145,15 @@ NamedIntrinsic Compiler::lookupHWIntrinsic(const char* methodName, InstructionSe
             }
         }
+
+        if ((result > NI_SW_INTRINSIC_START) && (result < NI_HW_INTRINSIC_END))
```
Shouldn't this be `result < NI_SW_INTRINSIC_END`

4creators replied:
Yes, it sounds better, but then I would add NI_HW_INTRINSIC_END with value equal to NI_SW_INTRINSIC_START

tannergooding replied:
I think they should be uniquely defined values to help prevent confusion

tannergooding commented on src/jit/namedintrinsiclist.h:
```diff
@@ -20,6 +20,10 @@ enum NamedIntrinsic : unsigned int
 #define HARDWARE_INTRINSIC(id, name, isa, ival, size, numarg, t1, t2, t3, t4, t5, t6, t7, t8, t9, t10, category, flag) \
     NI_##id,
 #include "hwintrinsiclistxarch.h"
+    NI_SW_INTRINSIC_START,
```
We should have `NI_HW_INTRINSIC_END` before this, to clearly define the separate ranges.

4creators replied:
Agrred, I have ended up with same conclusion above 😄

tannergooding commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -1023,47 +1024,96 @@ public static class Sse2
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static Vector128<uint> SetAllVector128(uint value)
+        {
+            return Shuffle(ConvertScalarToVector128UInt32(value), 0);
```
could you break these lines apart, with temps and comments.

The JIT should be able to properly optimize them, and it makes it easier to read for us.

4creators replied:
Will do

tannergooding commented on tests/src/JIT/HardwareIntrinsics/X86/Sse2/ConvertScalarToVector128Int64.cs:
```diff
@@ -22,16 +22,16 @@ static unsafe int Main(string[] args)
             if (Sse2.IsSupported)
             {
-                using (TestTable<long> longTable = new TestTable<long>(new long[2] { 1, -5 }, new long[2]))
+                using (TestTable<long> longTable = new TestTable<long>(new long[2] { 1, -758765098768743212 }, new long[2]))
```
Unrelated test changes?

4creators replied:
Yeah, it slipped from other PR work while changing branches. Will remove.

4creators commented:
@tannergooding I will wait with changes until #16817 is merged since I need to rebase on top of it and update implementation here to handle `SetVector128`

4creators commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

4creators commented:
@dotnet-bot test Windows_NT x86 Release Innerloop Build and Test

fiigii commented on src/jit/CMakeLists.txt:
```diff
@@ -126,6 +126,7 @@ if (WIN32)
     hashbv.h
     host.h
     hostallocator.h
+    hwintrinsichelperlistxarch.h
```
I do not think a new list is necessary. Just use the original table and change `HW_Flag_NoCodeGen` intrinsics to return `nullptr` is enough.

4creators replied:
> I do not think a new list is necessary. Just use the original table and change HW_Flag_NoCodeGen intrinsics to return nullptr is enough.

Compiler decides which compilation path to chose based on return value from `NamedIntrinsic Compiler::lookupHWIntrinsic(const char* methodName, InstructionSet isa)` call which should be NI_Illegal. There is no point in going through HW intrinsic codegen at all, as managed implementation of intrinsic can be treated as normal function call.

fiigii replied:
> Compiler decides which compilation path to chose based on return value from NamedIntrinsic Compiler::lookupHWIntrinsic(const char* methodName, InstructionSet isa) call which should be NI_Illegal. 

If you want to cut off the intrinsic-recognition/importation earlier, I think just removing the definition of the managed-implemented helper from this table is okay.

If you want to explicitly treat these helpers in JIT, I prefer to return `nullptr` from the importer. 
Another table is unnecessary.

CarolEidt replied:
I agree that another table is unnecessary.

4creators replied:
> If you want to cut off the intrinsic-recognition/importation earlier, I think just removing the definition of the managed-implemented helper from this table is okay.

The point is that the way it was done at the beginning - removing definition from table - was deemed OK but the desire expressed both by @tannergooding and @CarolEidt was to have some reasonable way to control the correctness of this path via assertion.

see discussion here https://github.com/dotnet/coreclr/pull/16797#issuecomment-371172663  and here https://github.com/dotnet/coreclr/pull/16797#issuecomment-371294081

There is still one more way to do it without introducing additional complexity and new table and I will create separate branch to allow CR of that option

tannergooding replied:
> but the desire expressed both by @tannergooding and @CarolEidt was to have some reasonable way to control the correctness of this path via assertion.

A simple flag `HW_Flag_ManagedImplementation` (or something with a better name) should be completely sufficient for this.

fiigii replied:
> The point is that the way it was done at the beginning - removing definition from table - was deemed OK but the desire expressed both by @tannergooding and @CarolEidt was to have some reasonable way to control the correctness of this path via assertion.

That is another story. As I said, there are two solutions:
1. Managed-implemented helpers cannot reach the importer of HW intrinsics if the definition is removed from the table.

2. Checking the flag of `HW_Flag_NoCodeGen` (or another new name) and returning `nullptr` can solve the "control the correctness of this path via assertion." that @tannergooding and @CarolEidt mentioned.

fiigii replied:
BTW, I prefer that "Checking the flag of HW_Flag_NoCodeGen (or another new name) and returning nullptr", which let the managed-implemented helpers can leverage `NotSupported` checking (e.g., invalid type-argument for generic intrinsic) and managed code does not need to care about throwing the exception.

fiigii replied:
You used the first solution (removing the definition) in https://github.com/dotnet/coreclr/pull/16817. Please unify.

4creators replied:
That is just obvious and was indicated several times that #16817 has to be merged first to work on final solution here which will unfiy both and this is what I am doing here - the code in current version unified all implementations and the code in new version  will unify them as well.

fiigii replied:
> That is just obvious and was indicated several times

> the code in current version unified all implementations and the code in new version will unify them as well.

Sounds good. 
I was in traveling recently, so that I missed some conversation contexts. Sorry to duplicate the discussion.

fiigii commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -117,29 +120,40 @@ InstructionSet Compiler::lookupHWIntrinsicISA(const char* className)
+        if ((result > NI_SW_INTRINSIC_START) && (result < NI_SW_INTRINSIC_END))
+        {
+            return NI_Illegal;
+        }
```
This design changes the framework too much. I prefer to use `HARDWARE_INTRINSIC` with `HW_Flag_NoCodeGen` in the original table.

tannergooding replied:
Yeah, I wonder if defining a new group is the right thing to be doing.

It seems like a flag would be sufficient. Although we may want something that differentiates `NoCodeGen` and `Implemented in Managed Code`.

fiigii replied:
Agree, we can change the name of `HW_Flag_NoCodeGen`.

fiigii commented on src/mscorlib/src/System/Runtime/Intrinsics/X86/Sse2.cs:
```diff
@@ -968,6 +968,149 @@ public static class Sse2
+            // Create { -- -- -- -- -- -- vl vl }
+            Vector128<int> tmpVector = Sse.StaticCast<short, int>(UnpackLow(vector, vector));
+            // Create { vl vl vl vl vl vl vl vl } and return result
+            return Sse.StaticCast<int, short>(Shuffle(tmpVector, (byte)0));
```
Please use `Avx2.BroadcastScalarToVector128` when AVX2 is available for all the helpers.

tannergooding replied:
I would prefer the changes to go in "as-as" (just Sse/Sse2 codegen) and for a later change to improve the codegen (with AVX support, for example).

4creators commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

4creators commented:
Addressed code review feedback. 

@CarolEidt @fiigii @tannergooding PTAL

fiigii commented on src/coreclr/hosts/corerun/corerun.cpp:
```diff
@@ -33,7 +33,7 @@ static const wchar_t *coreCLRDll = W("CoreCLR.dll");
 
 // Encapsulates the environment that CoreCLR will run in, including the TPALIST
-class HostEnvironment 
+class HostEnvironment
```
Merge conflict?

4creators replied:
This file and `Directory.Build.targets` are part of local working infrastructure which slipped into commit. Removing.

fiigii commented on Directory.Build.targets:
```diff
@@ -0,0 +1,46 @@
+        DestinationFiles="$(MSBuildThisFileDirectory)bin\tests\$(OS).$(PlatformName).$(Configuration)\Tests\Core_Root\$(TargetName).pdb"/>
+
+  </Target>
+</Project>
```
Why do you add this file?

4creators commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

fiigii commented on src/jit/hwintrinsiclistxarch.h:
```diff
@@ -435,6 +434,34 @@ HARDWARE_INTRINSIC(PCLMULQDQ_IsSupported,                            "get_IsSupp
+//
+//  End of software helper intrinsics
+//
+HARDWARE_INTRINSIC(SW_INTRINSIC_END,                                 "SW_INTRINSIC_END",                                 POPCNT,     -1,           0,            0,           {INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid,   INS_invalid},           HW_Category_SimpleSIMD,                        HW_Flag_NoFlag)
```
I do not think another group in the table is necessary, please just giving them a flag.

fiigii commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -137,9 +140,18 @@ NamedIntrinsic Compiler::lookupHWIntrinsic(const char* methodName, InstructionSe
+
+        if ((NI_SW_INTRINSIC_START < result) && (result < NI_SW_INTRINSIC_END))
+        {
+            return NI_Illegal;
```
Unnecessary.

4creators replied:
IMO having a new flag and going through intrinsic recognition deep up to the point where we return nullptr after checking flag creates longer and less efficient codepath. Can you explain what type of benefits will we have by implementing your proposal and what is difference in code path between tthis two solution and why your poposal is more efficient?

fiigii replied:
> checking flag creates longer and less efficient codepath. 

I do not think there is any issue of efficiency. Clear code is much more important.

> explain what type of benefits will we have by implementing your proposal

1. Managed-implemented intrinsics also can use the code of throwing `PlatformNotSupported` or type `NotSupported`.
2. We will move certain implementations to CodGen in the future.
3. A new range `SW_INTRINSIC_START\SW_INTRINSIC_END` is not necessary.

In summary, I just want to change the code as less as possible.

fiigii replied:
> flag creates longer and less efficient codepath.

As I said, if you want to cut off the intrinsic-recognition/importation earlier, I think just removing the definition of the managed-implemented helper from this table is okay.

4creators replied:
Code where we will be returning `NI_Illegal` or `nullptr` is inside `Compiler::impIntrinsic` in importer.cpp:

```C++
    if ((methodFlags & CORINFO_FLG_JIT_INTRINSIC) != 0)
    {
        // The recursive calls to Jit intrinsics are must-expand by convention.
        mustExpand = mustExpand || gtIsRecursiveCall(method);

        if (intrinsicID == CORINFO_INTRINSIC_Illegal)
        {
            ni = lookupNamedIntrinsic(method);

#ifdef FEATURE_HW_INTRINSICS
            if (ni > NI_HW_INTRINSIC_START && ni < NI_HW_INTRINSIC_END)
            {
                return impHWIntrinsic(ni, method, sig, mustExpand);
            }
#endif // FEATURE_HW_INTRINSICS
        }
    }
```

fiigii replied:
If an intrinsic is not in the table, `lookupHWIntrinsic` will automatically return `NI_Illegal` for it. 
Do not need any change.

4creators replied:
We can modify `Compiler::impIntrinsic` to get even more efficient:

```C++
        if (intrinsicID == CORINFO_INTRINSIC_Illegal)
        {
            ni = lookupNamedIntrinsic(method);

#ifdef FEATURE_HW_INTRINSICS
            if (ni > NI_HW_INTRINSIC_START && ni < NI_SW_INTRINSIC_START)
            {
                return impHWIntrinsic(ni, method, sig, mustExpand);
            }
            else if (ni > NI_SW_INTRINSIC_START && ni < NI_SW_INTRINSIC_END)
            {
                return nullptr;
            }
#endif // FEATURE_HW_INTRINSICS
        }
```

fiigii replied:
> We can modify Compiler::impIntrinsic to get even more efficient:

I do not think so.

4creators replied:
See example: https://github.com/dotnet/coreclr/compare/master...dotnetrt:SetAllVector128v3

4creators replied:
> > We can modify Compiler::impIntrinsic to get even more efficient:

> I do not think so.

Can you explain?

4creators replied:
> If an intrinsic is not in the table, lookupHWIntrinsic will automatically return NI_Illegal for it.
Do not need any change.

But than you would support any intrinsic which is not in the table without any way to assert it is correct to do so.

fiigii replied:
If an intrinsic is not in the table, lookupHWIntrinsic will automatically return NI_Illegal for it. That is exactly same with what you are doing now, but no change required.

4creators replied:
> I do not think another group in the table is necessary, please just giving them a flag and checking the flag to return `nullptr` in `Compiler::impHWIntrinsic`.

Furthermore, how would you check the flag on intrinsic which is not in the table?

fiigii replied:
To be clear, there are two solutions:
1. Remove the definition of managed-implemented intrinsic from the table, and no any other change required 
2. Checking flag to return nullptr, which needs the definition in the table.

tannergooding replied:
I would say that 2 is the better route overall, as it allows us to have explicit asserts/checks for these intrinsics elsewhere, if required.

@CarolEidt, do you have a preference?

4creators replied:
> To be clear, there are two solutions:

Why there are only 2 solutions available? Is it an objective, proven mathematically fact or it is just your opinion? If it's the second than it is up for discussion ...

4creators replied:
> I would say that 2 is the better route overall, as it allows us to have explicit asserts/checks for these intrinsics elsewhere, if required.

I am not a fan of flags usage since they are:

1. Not very well defined and not documented in a way which helps in using them.
2, Require additional check to use them.

Furthermore, the code path used in current solution simply treats managed implementation as ordinary managed function with `AggresiveInlining` attribute and inlines it.

There are no obvious needs for any further checks therefore implementing flag based solution for unclear future usage is an overengineering.

4creators replied:
IMO current 3rd solution allows to:

1. have positive asserts on all supported intrinsics.
2. saves 2 function calls
3. is more clear to follow overall for code users who have not created code (current preferences may be based on natural bias on using known, existing solution despite that it is more complex).

CarolEidt replied:
I think we should simply remove these from the table and have it "just work". Anything else seems like over-engineering to me. I don't think any of the other alternatives are as clean. If they are in the table with a flag, IMO it pollutes the space of what are expected to be actual intrinsics (i.e. methods for which the *JIT* generates code directly). If they are in their own table, it begs the question of why they're there at all. They're not really intrinics.

If we find a need for something more complex down the road we can always add it.

4creators replied:
@CarolEidt Thanks for clear guidelines.

4creators commented:
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

4creators commented:
Addressed all CR feedback, CI is green, 

@CarolEidt @fiigii @tannergooding PTAL

CarolEidt commented on src/jit/hwintrinsicxarch.cpp:
```diff
@@ -124,7 +124,7 @@ InstructionSet Compiler::lookupHWIntrinsicISA(const char* className)
 //
 // Return Value:
-//    Id for the hardware intrinsic.
+//    Id for the hardware intrinsic
```
Not sure why this change was made. Although many comments lack a '.' the function guidelines illustrate their use in this context.

4creators replied:
This is result of rolling back previous changes to return to original state, what seems was not as precise as it should be.

CarolEidt commented on src/jit/namedintrinsiclist.h:
```diff
@@ -27,8 +27,8 @@ enum NamedIntrinsic : unsigned int
 #endif // !defined(_TARGET_XARCH_) && !defined(_TARGET_ARM64_)
-    NI_HW_INTRINSIC_END
-#endif
+    NI_HW_INTRINSIC_END,
```
Another gratuitous change.

4creators replied:
The same has happened here

4creators replied:
Will fix them before merge

4creators commented:
@fiigii @tannergooding Could you pls give a final review pass on this PR
